### PR TITLE
Struct Method Renaming

### PR DESCRIPTION
**Pull request by bguesman**
Implements basic renaming support for methods in structs.

### Strategy

In HLSL, struct methods can reference struct members and other struct methods in their bodies, regardless of the order they were declared in. I.e., the following is valid:

```
struct {
  int A() {
    // A() can reference B() and c before they are declared, in other words 
    // "out of order"
    return B() + c;
  }

  int B() {
    return 4;
  }

  int c;
};
```

To allow for this, we take the following approach when renaming struct members:
- First, rename all struct member declarations (methods and member variables), adding their renaming info to the top level environment `env`.
- Then, rename all method bodies. At this point all members will be renamed, so their new names will be resolved correctly.

One slight nuance: we want expressions evaluated in member bodies to know about renamed members. Rather than change the way the expression renamer works to search the member renames table, we temporarily augment identifier renames table with the member renames, only in the context of renaming the function body. Here's the relevant snippet:

```
let methodEnv = env.onEnterScope env stmt
let methodEnv = {methodEnv with identRenames = Map.fold (fun acc k v -> Map.add k v acc) methodEnv.identRenames methodEnv.memberRenames}
renStmt methodEnv stmt |> ignore<Env>
env
```

### Limitations

This will fail under certain pathological circumstances. For instance:

```
struct A {
  int a;
}

struct B {
  int b() {
    return a;
  }
}
```

Will recognize the `a` referenced in `b()` as the `a` in `A`, because b will have access to all member name replacements across all structs processed so far. The correct behavior is, of course, to recognize that `a` is not defined within `b()`'s scope and leave it as-is.

There are other corner cases like this one. Implementing this to full correctness would require a form of type inference to disambiguate between symbols from different user-defined types.

### CODE REVIEW

File: tests/unit/out_of_order_struct_member_functions.hlsl [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    float my_other_function() {
+        int something = 6;
+        return 10 + member_var + something;
+    }
+
+    int member_var;
+};
+
+float my_other_function() {
+    return 10 + member_var;
```
Comment (laurentlb) [inline tests/unit/out_of_order_struct_member_functions.hlsl:15]: Is it allowed to call `member_var` here? I'm not familiar with HLSL, but in most other languages, I'd expect to have `somestruct.member_var`.
Can you also call `my_function` from here?

Comment (laurentlb) [reply tests/unit/out_of_order_struct_member_functions.hlsl:15]: Alright, but if you call `MyStruct.my_function` (outside the struct definition), will it get renamed?
I'd like to make sure the renaming doesn't break the shader.

Comment (bguesman) [reply tests/unit/out_of_order_struct_member_functions.hlsl:15]: I believe it should, since the dot operator trick that checks the members list in the environment is still in there. But I should add that to this test—will do when I'm back at my PC in a week.
